contracts/exchange/contracts/src/MixinMatchOrders.sol

```diff
@@ -262,6 +355,237 @@ contract MixinMatchOrders is
+        }
+        if (rightOrders.length == 0) {
+            _rrevert(BatchMatchOrdersError(BatchMatchOrdersErrorCodes.ZERO_RIGHT_ORDERS));
+        }
+
+        // Ensure that the left and right arrays are compatible.
+        if (leftOrders.length != leftSignatures.length) {
+            _rrevert(BatchMatchOrdersError(BatchMatchOrdersErrorCodes.INCOMPATIBLE_LEFT_ORDERS));
+        }
+        if (rightOrders.length != rightSignatures.length) {
+            _rrevert(BatchMatchOrdersError(BatchMatchOrdersErrorCodes.INCOMPATIBLE_RIGHT_ORDERS));
+        }
+
+        // Without simulating all of the order matching, this program cannot know how many
+        // matches there will be. To ensure that batchMatchedFillResults has enough memory
+        // allocated for the left and the right side, we will allocate enough space for the
+        // maximum amount of matches (the left side length added to the right side length).
+        uint256 maxLength = leftOrders.length + rightOrders.length;
+        batchMatchedFillResults.left = new LibFillResults.FillResults[](maxLength);
+        batchMatchedFillResults.right = new LibFillResults.FillResults[](maxLength);
+
+        // Set up initial indices.
+        uint256 matchCount;
+        uint256 leftIdx = 0;
+        uint256 rightIdx = 0;
+
+        // Keep local variables for orders, order info, and signatures for efficiency.
+        LibOrder.Order memory leftOrder = leftOrders[0];
+        LibOrder.Order memory rightOrder = rightOrders[0];
+        LibOrder.OrderInfo memory leftOrderInfo = getOrderInfo(leftOrder);
+        LibOrder.OrderInfo memory rightOrderInfo = getOrderInfo(rightOrder);
+
+        // Loop infinitely (until broken inside of the loop), but keep a counter of how
+        // many orders have been matched.
+        for (matchCount = 0;;) {
+            // Match the two orders that are pointed to by the left and right indices
+            LibFillResults.MatchedFillResults memory matchResults = _matchOrders(
+                leftOrder,
+                rightOrder,
+                leftSignatures[leftIdx],
+                rightSignatures[rightIdx],
+                shouldMaximallyFillOrders
+            );
+
+            // Update the orderInfo structs with the updated takerAssetFilledAmount
+            leftOrderInfo.orderTakerAssetFilledAmount = _safeAdd(
+                leftOrderInfo.orderTakerAssetFilledAmount,
+                matchResults.left.takerAssetFilledAmount
+            );
+            rightOrderInfo.orderTakerAssetFilledAmount = _safeAdd(
+                rightOrderInfo.orderTakerAssetFilledAmount,
+                matchResults.right.takerAssetFilledAmount
+            );
+
+            // Add the matchResults and the profit made during the match to the
+            // batchMatchedFillResults for this batch.
+            batchMatchedFillResults.left[matchCount] = matchResults.left;
```

Comment:
These results correspond to the index of each match rather than the index of each order. I feel like the latter would be more useful, since you can quickly check the results to see how much of each order has been filled (i.e `matchedFillResults.left[3]` would tell you how much of `leftOrders[3]` has been filled). It would probably require us to track a `leftFillResults` and `rightFillResults`, much in the same way we do for `orderInfo` in this implementation.